tests/EZConfig.hs

```diff
@@ -30,15 +44,59 @@ spec = do
     it "Fails on the non-existent key M-10" $
       readKeySequence def "M-10" `shouldBe` Nothing
 
-regularKeys :: ([String], [KeySym])
-regularKeys = unzip . map (first (: ""))
-            $ zip ['!'    .. '~'   ] [xK_exclam       .. xK_asciitilde]
-           ++ zip ['\xa0' .. '\xff'] [xK_nobreakspace .. xK_ydiaeresis]
+-- | Parsing preserves all info that printing does.
+prop_encodePreservation :: KeyString -> Property
+prop_encodePreservation (coerce -> s) = parse s === (parse . pp =<< parse s)
+ where parse = runParser (parseKeySequence def)
+       pp    = unwords . map keyToString
+
+-- | Printing preserves all info that parsing does.
+prop_decodePreservation :: NonEmptyList (AKeyMask, AKeySym) -> Property
+prop_decodePreservation (getNonEmpty >>> coerce -> xs) =
+  Just (pp xs) === (fmap pp . parse $ pp xs)
+ where parse = runParser (parseKeySequence def)
+       pp    = unwords . map keyToString
 
 -- | QuickCheck can handle the 8! combinations just fine.
 modifiers :: [String]
-modifiers = map concat $
-  permutations ["M-", "C-", "S-", "M1-", "M2-", "M3-", "M4-", "M5-"]
+modifiers = map concat $ permutations mods
+
+mods :: [String]
+mods = ["M-", "C-", "S-", "M1-", "M2-", "M3-", "M4-", "M5-"]
```

Comment:
Neither of the tested props detects that `M-` isn't preserved in `pp ∘ p`. Is there a reason why you don't test `pp ∘ p ≡ id` and `p ∘ pp ≡ id` instead?

Reply:
For one, the pretty-printer, being a pure function, does not know about the users mod key and so prints `M-` as whatever mod key it actually is (`M1-`, `M2-`, ...).

The more "fundamental" reason is that we allow permutations of the modifiers, but the pretty-printer arranges things deterministically.  So for example `p ∘ pp ∘ p "S-M-a" ≡ p "S-M-a"` but `pp ∘ p "S-M-a" = "M-S-a" ≠ "S-M-a"`.

Reply:
Oh, I missed that the list of mods was already used by existing tests, so leaving out `M-` wasn't an option. Now it makes sense that `M-` is here even though for the new tests it's not useful.